### PR TITLE
feat: Add acyclic and tred support

### LINKED ISSUE #210: Transitive reduction filter for directed graphs - Graphviz
Hello, I'm using d3-graphviz to draw a directed graph.

I need to apply [tred](https://graphviz.org/pdf/tred.1.pdf). Is it possible with this library? If not, how can I apply this algorithm?

### PR DESCRIPTION
**Pull request by GordonSmith**
Fixes #210

### CODE REVIEW

File: src-cpp/graphviz/main.cpp [hunk 1]
```diff
@@ -137,122 +137,67 @@ const char *Graphviz::layout(const char *src, const char *format, const char *en
...
+    acyclic_outFile_buff = "";
+    acyclic_outFile = acyclic_outFile_buff.c_str();
+    acyclic_num_rev = 0;
+    bool retVal = false;
+    Agraph_t *g = agmemread(src);
+    if (g)
     {
-        if (agtail(e) != aghead(e))
-            rv++;
+        FILE *outFile = fopen("tmp.dot", "w");
```
Comment (Smattr) [inline src-cpp/graphviz/main.cpp:150]: Do you not want to (1) use a `getenv("TMPDIR") || "/tmp"` prefix, (2) use `mkstemp` or something to ensure parallel executions won’t interfere with each other, and (2) check for failure here?

Alternatively to using a temporary file, if this only runs on !Windows, you could do this entirely in memory with `open_memstream`.

File: src-ts/graphviz.ts [hunk 2]
```diff
@@ -155,24 +156,90 @@ export class Graphviz {
         return retVal;
     }
 
+    /**
+      * acyclic is a filter that takes a directed graph as input and outputs a copy of the graph with sufficient edges reversed to make the graph acyclic. The reversed edge inherits all of the attributes of the original edge. The optional file argument specifies where the the input graph is stored; by default.
```
Comment (Smattr) [inline src-ts/graphviz.ts:160]: nit: s/the the/the

File: src-ts/graphviz.ts [hunk 3]
```diff
@@ -155,24 +156,90 @@ export class Graphviz {
...
+      * 
+      * @param dotSource Required - graph definition in [DOT](https://graphviz.gitlab.io/doc/info/lang.html) language
+      * @param doWrite Enable output is produced, though the return value will indicate whether the graph is acyclic or not.
+      * @param verbose Print information about whether the file is acyclic, has a cycle or is undirected.
+      * @returns `{ acyclic: boolean, num_rev: number, outFile: string }` Boolean will be true if a cycle was found,  as well as.
```
Comment (Smattr) [inline src-ts/graphviz.ts:165]: nit: double space. Though it actually looks like there are more words missing here.

File: src-cpp/graphviz/main.cpp (hunk 1 above)
Comment (GordonSmith) [reply src-cpp/graphviz/main.cpp:150]: FYI neither of these are really a concern in WebAssembly as the files end up in memory anyway (and parallel execution will only happen with separate instances).
I will tidy up into a convenience class anyway as this is code is getting reproduced a few times now...

File: src-cpp/graphviz/main.cpp [hunk 4]
```diff
@@ -86,182 +89,88 @@ void Graphviz::createFile(const char *path, const char *data)
...
 {
-    return agdegree(n->root, n, TRUE, FALSE);
-}
-
-/* need outdegree without selfarcs */
-int myoutdegree(Agnode_t *n)
-{
-    Agedge_t *e;
-    int rv = 0;
+    lastErrorStr[0] = '\0';
```
Comment (GordonSmith) [inline src-cpp/graphviz/main.cpp:121]: Should be moved to constructor

File: src-cpp/graphviz/util.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+class TempFileBuffer
+{
+private:
+    FILE *filePointer = nullptr;
+    char tempFileName[L_tmpnam];
+
+public:
+    TempFileBuffer();
+    ~TempFileBuffer();
```
Comment (Smattr) [inline src-cpp/graphviz/util.hpp:24]: You probably want to `= delete` the copy/move constructor/assignment for this class, right?

File: src-cpp/graphviz/util.hpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <string>
+
+class StringBuffer
```
Comment (Smattr) [inline src-cpp/graphviz/util.hpp:5]: Not sure I understand the intent of this class. Why not just use a `std::string`?

File: src-cpp/graphviz/util.hpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+public:
+    operator std::string() const;
+    operator const char *() const;
+    StringBuffer &operator=(const std::string &str);
+};
+
+class TempFileBuffer
+{
+private:
+    FILE *filePointer = nullptr;
```
Comment (Smattr) [inline src-cpp/graphviz/util.hpp:19]: `#include <cstdio>`?

File: src-cpp/graphviz/main.hpp [hunk 8]
```diff
@@ -1,10 +1,9 @@
 #include <string>
+#include "util.hpp"
 
 class Graphviz
 {
 private:
```
Comment (Smattr) [inline src-cpp/graphviz/main.hpp:6]: nit: I think you could also remove this line.

File: src-cpp/graphviz/util.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    return filePointer;
+}
+
+TempFileBuffer::operator std::string() const
+{
+    std::string content;
+    if (filePointer != nullptr)
+    {
+        std::rewind(filePointer);
+        char buffer[256];
```
Comment (Smattr) [inline src-cpp/graphviz/util.cpp:49]: nit: maybe you want to use `BUFSIZ` for the size here.

**Review by Smattr — COMMENTED** [review]
I’m not an expert in most of this stuff, but in general LGTM!

File: src-cpp/graphviz/util.hpp (hunk 6 above)
Comment (GordonSmith) [reply src-cpp/graphviz/util.hpp:5]: Might be a naming issue (StringAdapter might be better), but this allows the IDL generated "glue" code to return a `const char *` directly, without having to declare an explicit `const char *` member.

File: src-cpp/graphviz/util.cpp (hunk 9 above)
Comment (GordonSmith) [reply src-cpp/graphviz/util.cpp:49]: Should be able to remove the buff and loop altogether...

File: src-cpp/graphviz/util.hpp (hunk 5 above)
Comment (GordonSmith) [reply src-cpp/graphviz/util.hpp:24]: True.

**Pull request merged**